packages/block-library/src/page-list/index.php

```diff
@@ -0,0 +1,216 @@
+ *
+ * @param  array $context Navigation block context.
+ * @return array Font size CSS classes and inline styles.
+ */
+function block_core_page_list_build_css_font_sizes( $context ) {
+	// CSS classes.
+	$font_sizes = array(
+		'css_classes'   => array(),
+		'inline_styles' => '',
+	);
+
+	$has_named_font_size  = array_key_exists( 'fontSize', $context );
+	$has_custom_font_size = array_key_exists( 'customFontSize', $context );
+
+	if ( $has_named_font_size ) {
+		// Add the font size class.
+		$font_sizes['css_classes'][] = sprintf( 'has-%s-font-size', $context['fontSize'] );
+	} elseif ( $has_custom_font_size ) {
+		// Add the custom font size inline style.
+		$font_sizes['inline_styles'] = sprintf( 'font-size: %spx;', $context['customFontSize'] );
+	}
+
+	return $font_sizes;
+}
+
+/**
+ * Outputs Page list markup from an array of pages with nested children.
+ *
+ * @param array $nested_pages The array of nested pages.
+ *
+ * @return string List markup.
+ */
+function render_nested_page_list( $nested_pages ) {
+	if ( empty( $nested_pages ) ) {
+		return;
+	}
+	$markup = '';
+	foreach ( (array) $nested_pages as $page ) {
+		$css_class = array( 'wp-block-pages-list__item' );
+		if ( isset( $page['children'] ) ) {
+			$css_class[] = 'has-child';
+		}
+		$markup .= '<li class="' . implode( ' ', $css_class ) . '">';
+		$markup .= '<a class="wp-block-pages-list__item__link" href="' . $page['link'] . '">' . $page['title'] . '</a>';
+		if ( isset( $page['children'] ) ) {
+			$markup .= '<span class="wp-block-page-list__submenu-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 24 24" transform="rotate(90)"><path d="M8 5v14l11-7z"></path><path d="M0 0h24v24H0z" fill="none"></path></svg></span>';
+			$markup .= '<ul class="submenu-container">' . render_nested_page_list( $page['children'] ) . '</ul>';
+		}
+		$markup .= '</li>';
+	}
+	return $markup;
+}
+
+/**
+ * Outputs nested array of pages
+ *
+ * @param array $current_level The level being iterated through.
+ * @param array $children The children grouped by parent post ID.
+ *
+ * @return array The nested array of pages.
+ */
+function nest_pages( $current_level, $children ) {
+	if ( empty( $current_level ) ) {
+		return;
+	}
+	$nested_array = array();
+	foreach ( (array) $current_level as $key => $current ) {
+		if ( isset( $children[ $key ] ) ) {
+			$current['children'] = nest_pages( $children[ $key ], $children );
+		}
+		$nested_array[] = $current;
+	}
+	return $nested_array;
+}
+
+	/**
+	 * Renders the `core/page-list` block on server.
+	 *
+	 * @param array $attributes The block attributes.
+	 * @param array $content The saved content.
+	 * @param array $block The parsed block.
+	 *
+	 * @return string Returns the page list markup.
+	 */
+function render_block_core_page_list( $attributes, $content, $block ) {
+	static $block_id = 0;
+	$block_id++;
+
+	$all_pages = get_pages();
```

Comment:
Yeah these are really objects, not arrays as we understand them in JS 😅 

I'm not sure what the page order setting is supposed to do, apart from changing the order of display in `wp-admin/edit.php?post_type=page`. Trying to add all pages to a classic menu, the numbered ones are sitting between `0`s too 🤷‍♀️ 

I'd expect `get_pages` to offer a way of sorting them by order but it doesn't seem to, and I'm following the default order `get_pages` returns them in (apparently ASC).

Reply:
The pages widget does sort in the right order though, so probably good to replicate that, and it seems like an improvement over the old menus.

For get_pages, the `sort_column` argument should make sure the order is used for sorting. It's a shame there's no way to specify a secondary sort.

```php
get_pages( array( sort_column => 'menu_order' ) );
```

Reply:
I did some more investigation into this, as I became intrigued by the `hierarchical` argument of `get_pages`, which is always `true`.

It looks like if you do `get_pages( array( 'sort_column' => 'menu_order' ) )`, it by default returns all the pages in an order that takes account the nesting. The return value is a flat array, but the nested items are in the correct position. It looks like it should be possible to simplify how the nesting works in this block.

Reply:
I'm not sure there's much simplification can be done, as we still need to nest child pages under parents to generate the proper markup, so we need the pages to be indexed by their ID, which they're not when output by `get_pages`. I can't think of an easier way of doing this, but suggestions welcome!

Reply:
Yeah, it is still pretty tricky. I expected `hierarchical` to build a nested array much like you've done in this PR. I think 
 to render in this format, the rendering loop would have to walk ahead to the next items to determine if they're children, and then add the nesting markup when that's the case.

Does feel a bit like an unnecessary refactoring given what's in this PR works, so maybe not worth it.

It could still be a decent optimisation to set hierarchical option to false in the pages block, as `get_pages` does a whole bunch of work that's not needed:
- https://github.com/WordPress/wordpress-develop/blob/4cfbf629ca231820cc692049be30ef588101e523/src/wp-includes/post.php#L5852-L5854
- https://github.com/WordPress/wordpress-develop/blob/4cfbf629ca231820cc692049be30ef588101e523/src/wp-includes/post.php#L5453-L5480